src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -11,6 +11,316 @@ namespace System.Text.RegularExpressions
     /// <summary>Detects various forms of prefixes in the regular expression that can help FindFirstChars optimize its search.</summary>
     internal static class RegexPrefixAnalyzer
     {
+        /// <summary>Cache of ToString() strings for the ASCII chars.</summary>
+        /// <remarks>The strings are lazily created on first use.</remarks>
+        private static string[]? s_asciiCharStrings;
+
+        /// <summary>Gets the ToString() string for the specified char.</summary>
+        private static string GetCharString(char ch)
+        {
+            // If the character isn't ASCII, just ToString it.
+            if (ch >= 128)
+            {
+                return ch.ToString();
+            }
+
+            // Use a lazily-initialized cache of strings for ASCII characters. The overall cache is initialized
+            // with Interlocked.CompareExchange simply to avoid accidentally throwing out a lot of strings accidentally.
+
+            string[] asciiCharString =
+                s_asciiCharStrings ??
+                Interlocked.CompareExchange(ref s_asciiCharStrings, new string[128], null) ??
+                s_asciiCharStrings;
+
+            return asciiCharString[ch] ??= ch.ToString();
+        }
+
+        /// <summary>Finds an array of multiple prefixes that a node can begin with.</summary>
+        /// <param name="node">The node to search.</param>
+        /// <param name="ignoreCase">true to find ordinal ignore-case prefixes; false for case-sensitive.</param>
+        /// <returns>
+        /// If a fixed set of prefixes is found, such that a match for this node is guaranteed to begin
+        /// with one of those prefixes, an array of those prefixes is returned.  Otherwise, null.
+        /// </returns>
+        public static string[]? FindPrefixes(RegexNode node, bool ignoreCase)
+        {
+            // Minimum string length for prefixes to be useful. If any prefix has length 1,
+            // then we're generally better off just using IndexOfAny with chars.
+            const int MinPrefixLength = 2;
+
+            // Arbitrary string length limit (with some wiggle room) to avoid creating strings that are longer than is useful and consuming too much memory.
+            const int MaxPrefixLength = 8;
+
+            // Arbitrary limit on the number of prefixes to find. If we find more than this, we're likely to be spending too much time finding prefixes that won't be useful.
+            const int MaxPrefixes = 16;
```

Comment:
Curious as a follow-up, if we wanted to handle large alternation patterns (to make use of Aho–Corasick), would we use a simpler analysis (e.g. just looking for `Literal1|Literal2|...`)?
In that case, I'd assume we could use a much larger cutoff on the number of prefixes without the same worry about analysis time.

Reply:
> Curious as a follow-up, if we wanted to handle large alternation patterns (to make use of Aho–Corasick), would we use a simpler analysis (e.g. just looking for Literal1|Literal2|...)?

We'd have to rethink optimizations done earlier in the process. If you have an alternation like "abc|abd|acd", by the time it gets here it's been rewritten to be something more like "a(?:b[cd]|cd)".